core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/RootVariableSource.java

```diff
@@ -0,0 +1,246 @@
+
+                currentEntity = ConfigUtils.extractGenericTypeParameterOrFail(ShadowVariableUpdater.class.getSimpleName(),
+                        currentEntity,
+                        memberAccessor.getType(), memberAccessor.getGenericType(), ShadowVariableUpdater.class,
+                        memberAccessor.getName());
+
+                hasListMemberAccessor = true;
+            } else {
+                var memberAccessor = getMemberAccessor(currentEntity, pathPart,
+                        memberAccessorFactory, descriptorPolicy);
+
+                if (!hasListMemberAccessor) {
+                    listMemberAccessors.add(memberAccessor);
+                }
+
+                var isVariable = false;
+
+                for (var annotation : EntityDescriptor.getVariableAnnotationClasses()) {
+                    if (getAnnotation(currentEntity, memberAccessor.getName(), annotation) != null) {
+                        isVariable = true;
+                        break;
+                    }
+                }
+
+                chainToVariable.add(memberAccessor);
+                for (var chain : chainAfterVariable) {
+                    chain.add(memberAccessor);
+                }
+                if (isVariable) {
+                    List<MemberAccessor> chainAfter = new ArrayList<>();
+
+                    chainAfter.add(memberAccessor);
+                    chainAfterVariable.add(chainAfter);
+
+                    var chainBefore = new ArrayList<>(chainToVariable.subList(0, chainToVariable.size() - 1));
+                    variablePaths.add(new VariablePath(currentEntity, pathPart, chainBefore, chainAfter));
+                    afterVariable = true;
+                }
+                currentEntity = memberAccessor.getType();
+            }
+        }
+
+        @SuppressWarnings("unchecked")
+        BiConsumer<? extends Entity_, Consumer<Value_>> valueEntityFunction;
+
+        List<MemberAccessor> finalChainToVariable = chainToVariable.subList(0, chainToVariable.size() - 1);
+        if (!hasListMemberAccessor) {
+            valueEntityFunction = (entity, consumer) -> {
+                Object current = entity;
+                for (var accessor : finalChainToVariable) {
+                    current = accessor.executeGetter(current);
+                    if (current == null) {
+                        return;
+                    }
+                }
+                consumer.accept((Value_) current);
+            };
+        } else {
+            valueEntityFunction = (entity, consumer) -> {
+                Object current = entity;
+                for (var accessor : listMemberAccessors) {
+                    current = accessor.executeGetter(current);
+                    if (current == null) {
+                        return;
+                    }
+                }
+
+                Iterable<Object> iterable = (Iterable<Object>) current;
+                outer: for (var item : iterable) {
```

Comment:
Can we do this without labels? My experience has been that, whenever I felt like I need to use explicit goto, there was a better code structure that avoided it.